src/java/org/apache/cassandra/index/sai/disk/v2/sortedterms/SortedTermsWriter.java

```diff
@@ -0,0 +1,223 @@
+     * @param trieWriter where to write the trie that maps the terms to point ids
+     */
+    public SortedTermsWriter(@Nonnull IndexOutput termsData,
+                             @Nonnull IndexOutput termsDataBlockOffsets,
+                             @Nonnull IndexOutputWriter trieWriter)
+    {
+        this.trieWriter = new IncrementalDeepTrieWriterPageAware<>(trieSerializer, trieWriter.asSequentialWriter());
+        this.termsOutput = termsData;
+        this.bytesStartFP = termsData.getFilePointer();
+        this.offsetsOutput = termsDataBlockOffsets;
+
+        offsetsBuilder = PackedLongValues.deltaPackedBuilder(PackedInts.COMPACT);
+    }
+
+    /**
+     * Appends a term at the end of the sequence.
+     * Terms must be added in lexicographic order.
+     *
+     * @throws IOException if write to disk fails
+     * @throws IllegalArgumentException if the term is not greater than the previous added term
+     */
+    public void add(final @Nonnull ByteComparable term) throws IOException
+    {
+        tempTerm.clear();
+        copyBytes(term, tempTerm);
+
+        final BytesRef termRef = tempTerm.get();
+        final BytesRef prevTermRef = this.prevTerm.get();
+
+        Preconditions.checkArgument(prevTermRef.length == 0 || prevTermRef.compareTo(termRef) < 0,
+                                    "Terms must be added in lexicographic ascending order.");
+        writeTermData(termRef);
+        writeTermToTrie(term);
+
+        maxLength = Math.max(maxLength, termRef.length);
+        swapTempWithPrevious();
+        pointId++;
+    }
+
+    private void writeTermToTrie(ByteComparable term) throws IOException
+    {
+        trieWriter.add(term, pointId);
+    }
+
+    private void writeTermData(BytesRef term) throws IOException
+    {
+        if ((pointId & TERMS_DICT_BLOCK_MASK) == 0)
+        {
+            offsetsBuilder.add(termsOutput.getFilePointer() - bytesStartFP);
+
+            termsOutput.writeVInt(term.length);
+            termsOutput.writeBytes(term.bytes, term.offset, term.length);
+        }
+        else
+        {
+            final int prefixLength = StringHelper.bytesDifference(prevTerm.get(), term);
+            final int suffixLength = term.length - prefixLength;
+            assert suffixLength > 0: "terms must be unique";
+
+            termsOutput.writeByte((byte) (Math.min(prefixLength, 15) | (Math.min(15, suffixLength - 1) << 4)));
+            if (prefixLength >= 15)
+                termsOutput.writeVInt(prefixLength - 15);
+            if (suffixLength >= 16)
+                termsOutput.writeVInt(suffixLength - 16);
+
+            termsOutput.writeBytes(term.bytes, term.offset + prefixLength, term.length - prefixLength);
+        }
+    }
+
+    /**
+     * Flushes any in-memory buffers to the output streams.
+     * Does not close the output streams.
+     * No more writes are allowed.
+     */
+    public @Nonnull SortedTermsMeta finish() throws IOException
```

Comment:
The offsets are meant to be flushed to disk, which my original implementation did.  In the `SortedTermsReader` constructor, the whole bytes are on heap and this is incorrect.  

The reason the name of the class is `Direct` is because it is meant to be used, as it is in Lucene, directly from disk.